docs/rfc/rfc-1.adoc

```diff
@@ -0,0 +1,143 @@
+been released.
+
+=== Continuous Fees
+
+The system is able to collect fees in the form of tBTC when the BTC is
+deposited and when it is redeemed. Rather than paying operators at those two
+instances we would prefer to pay them continuously. Since we do not have the
+tBTC to pay them with until we collect the redemption fee (and can't mint it
+and maintain the peg), we must turn to another token. Instead, the system pays
+operators continuously in mintable T tokens and whenever it collects tBTC fees
+it uses those fees to buy T tokens back for its treasury.
+
+== In Depth
+
+=== Minting
+
+==== Deposit + Commitment
+
+Once we know the active wallet's public key hash, the dApp can put together a
+pay-to-script-hash (P2SH) address to receive the funds. This script will be
+unique to each depositor and will look like:
+
+```
+<eth-address> DROP
+<blinding-factor> DROP
+DUP HASH160 <signingGroupPubkey> EQUAL
+IF
+  CHECKSIG
+ELSE
+  DUP HASH160 <refundPubkey> EQUALVERIFY
+  <locktime> CHECKLOCKTIMEVERIFY DROP
+  CHECKSIG
+ENDIF
+```
+
+Since each depositor has their own ethereum address and their own secret
+blinding factor (which is an additional security layer), each depositor's
+script will be unique, and the hash of each depositor's script will be unique.
+
+In order to unlock the funds, one must provide the unhashed script, (which
+means that they know the eth address and blinding factor), as well as an
+unlocking script with a signature and public key. If the sig+pubkey matches the
+signing group public key, the funds are able to be moved immediately. If the
+sig+pubkey matches the refund public key, then the funds can be moved after 30
+days (specified as `locktime`).
+
+==== The Big Reveal
+
+After the deposit transaction has been mined, the user is able to reveal their
+ethereum address and blinding factor to the ethereum chain. The TBTC system listens
+for these sorts of messages and when it sees one, is able to generate a script that
+can spend the funds. Once successful, wallet moves the funds from the P2SH
+address to the main wallet address, which allows the system to mint tBTC and
+disables the 30-day refund.
+
+==== Automated Refunds
+
+A bitcoin transaction is an amount and a script. The script can be something as
+simple as "these funds can be spent by wallet 0xabc", or in our case, as
+complex as "these funds can be spent by wallet 0xabc but if they aren't spent
+within 30 days they can be spent by wallet 0x123". This gives us the ability to
+create deposits that automatically are refunded after 30 days if they aren't
+swept. Thus, if a user misfunds or they get cold feet (for any reason), all
+they need to do is not submit their reveal and wait 30 days.
+
+=== Redemption
+
+To initiate a redemption, a user supplies an amount `x` of TBTC and a bitcoin
+address. Then, the system calculates the redemption fee `fee`, and releases an
+amount of bitcoin `y` such that `x = y + fee` to the supplied bitcoin address.
```

Comment:
Hmm... I am not sure if we can assume `priceOf(1 TBTC) == priceOf(1 BTC)`. They should be ~ but what I think we observed for v1 is that TBTC price is a little higher given the deposit costs.

Reply:
Also, the amount of BTC released will probably be less given the BTC transfer fee - the redeemer should pay for it, it's the simplest solution I think. A separate question is how the system establishes BTC transaction fee. For v1, it has to be done on-chain. Are we willing to trust the judgment of clients? Having it on-chain feels more secure and we could avoid BTC transaction fee consensus issues between client instances.

Reply:
I'll clarify in the RFC, but the prices there aren't about the exchange rate - they're about maintaining the peg.

For example, if we're the custodians of 100 BTC, then we need to have minted 100 TBTC. If there's a 1% deposit fee, then 99 of TBTC go to the depositor, and 1 TBTC is sold off to buy T.

If someone comes back with 100 TBTC and there's a 1% redemption fee, then we'll burn 99 TBTC, release 99 BTC, and sell 1 TBTC to buy T.

Also, even though we're *releasing* 99 BTC in that case, that doesn't mean that 99 BTC lands in the hand of the redeemer, instead it's 99 minus the btc transaction fee.

> A separate question is how the system establishes BTC transaction fee. For v1, it has to be done on-chain.

Can you elaborate a bit here? I know that for liquidation recovery, we use the esplora fee API. How does v1 calculate the fee on-chain?

Reply:
Can't we let the redeemer choose what BTC transaction fee would be used? Probably with a governable minimum so people don't try to do zero-fee transactions, but they'd be otherwise free to decide how to split the amount of BTC that gets released.

Reply:
I now see we talk about two different things.

The RFC and @beaushinkle talk about redemption fee in TBTC, e.g. redeemer has to provide e.g. 1.1 TBTC to redeem 1 TBTC and 0.1 TBTC is used to buy back T and pay rewards to signers.

I, and @eth-r talk about Bitcoin transaction fee that needs to be paid for the redemption transaction and that needs to be chosen by someone. In tBTC v1 we let the redeemer to choose it and it could be increased by anyone (in practice, signers) in case the transaction could not be minted before the timeout, see [here](https://github.com/keep-network/tbtc/blob/fea1f4470544bc831cb789709eed6c027305cd48/solidity/contracts/deposit/Deposit.sol#L572).

I think I agree with @eth-r that it should be chosen by redeemer with some minimum. I think the minimum should be reasonable enough so that the wallet operators rarely have to increase it. Otherwise, we are requiring them to issue transaction just because the redeemer was too greedy and applied unrealistic constraints.

Side note, it could be a nice feature of the dApp to suggest the fee based on the current market conditions.